### PR TITLE
CB-5918 Add new plugin

### PR DESCRIPTION
**Pull request by SychevAndrey**

### CODE REVIEW

File: webapp/packages/plugin-holidays/package.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    "@cloudbeaver/plugin-top-app-bar": "^0",
+    "mobx": "^6",
+    "mobx-react-lite": "^4",
+    "react": "^18"
+  },
+  "peerDependencies": {},
+  "devDependencies": {
+    "@types/react": "*",
+    "typescript": "^5",
+    "typescript-plugin-css-modules": "*"
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/package.json]: please same versions as in other packages

File: webapp/packages/plugin-app-logo/package.json [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    "@cloudbeaver/core-routing": "^0",
+    "@cloudbeaver/core-sdk": "^0",
+    "@cloudbeaver/core-ui": "^0",
+    "@cloudbeaver/core-utils": "^0",
+    "@cloudbeaver/core-version": "^0",
+    "@cloudbeaver/core-view": "^0",
+    "@cloudbeaver/plugin-administration": "^0",
+    "@cloudbeaver/plugin-holidays": "^0",
+    "@cloudbeaver/plugin-settings-menu": "^0",
+    "@cloudbeaver/plugin-top-app-bar": "^0",
```
Comment (Wroud) [inline webapp/packages/plugin-app-logo/package.json]: seems like too many dependencies for "App logo"

File: webapp/packages/plugin-app-logo/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { appLogoPlugin } from './manifest.js';
+
+export * from './PluginBootstrap.js';
+
+export default appLogoPlugin;
```
Comment (Wroud) [inline webapp/packages/plugin-app-logo/src/index.ts:13]: please also export as named export:
```
export { appLogoPlugin }
```

File: webapp/packages/plugin-app-logo/src/PluginBootstrap.ts [hunk 4]
```diff
@@ -7,18 +7,22 @@
...
     private readonly wizardTopAppBarService: WizardTopAppBarService,
   ) {
     super();
   }
 
-  override register(): void | Promise<void> {
+  override register() {
+    this.topNavService.placeholder.add(Logo, 0);
     this.administrationTopAppBarService.placeholder.add(Logo, 0);
     this.wizardTopAppBarService.placeholder.add(Logo, 0);
```
Comment (Wroud) [inline webapp/packages/plugin-app-logo/src/PluginBootstrap.ts]: seems like you need to revert this change because `Logo` was added separately to the public and administration part on purpose

File: webapp/packages/plugin-holidays/package.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    "clean": "rimraf --glob dist",
+    "lint": "eslint ./src/ --ext .ts,.tsx",
+    "validate-dependencies": "core-cli-validate-dependencies",
+    "update-ts-references": "yarn run clean && typescript-resolve-references"
+  },
+  "dependencies": {
+    "@cloudbeaver/core-blocks": "^0",
+    "@cloudbeaver/core-di": "^0",
+    "@cloudbeaver/core-utils": "^0",
+    "@cloudbeaver/plugin-administration": "^0",
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/package.json]: In case this plugin depends on administration, you have two options:
Make two plugins:
`plugin-holidays`
`plugin-holidays-administration`

or make a single plugin
`plugin-holidays-administration`
(in that case, it should not affect the public part of the application)

This is because the administration part is optional as is `public` part, so we need this ability to turn off one of these parts

File: webapp/packages/plugin-holidays/src/Christmas/Christmas.ts [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    }
+  });
+
+  stop = action(() => {
+    window.removeEventListener('resize', this.onResize);
+    window.removeEventListener('mousemove', this.onMouseMove);
+    document.removeEventListener('mouseleave', this.onMouseLeave);
+    this.isSnowFalling = false;
+    clearTimeout(this.stopTimeoutId);
+  });
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/Christmas/Christmas.ts]: use simple functions instead cuz you already made these methods as actions in the constructor
also you actually don't need mobx actions here because the only mutation is `this.isSnowFalling = false;` it's discrete and simple

File: webapp/packages/plugin-holidays/src/Christmas/Christmas.ts [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+    this.ctx.restore();
+
+    const addAngle = 0.1 / shape + 0.02 * velY;
+    flake.angle = flake.angle > 0 ? flake.angle + addAngle : flake.angle - addAngle;
+  }
+
+  snow = (timestamp: number) => {
+    if (!this.ctx || !this.canvas) {
+      return;
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/Christmas/Christmas.ts]: please avoid using anonymous function as much as possible (it's impossible to debug)
prefer methods declaration instead

File: webapp/packages/plugin-holidays/src/HolidayActionButton.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+export const HolidayActionButton = observer(function HolidayActionButton() {
+  const { holiday } = useService(HolidaysService);
+
+  const handleButtonClick = () => {
+    if (holiday?.isCelebrating) {
+      holiday.stopCelebration();
+    } else {
+      holiday?.startCelebration();
+    }
+  };
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/HolidayActionButton.tsx]: same here please avoid anonymous functions

File: webapp/packages/plugin-holidays/src/HolidayActionButton.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  return holiday ? (
+    <button
+      className={s(holidayStyles, { holidayButton: true, buttonActive: holiday.isCelebrating })}
+      title={holiday.name}
+      onClick={handleButtonClick}
+    >
+      <img width={24} height={24} src={holiday.iconSrc} />
+    </button>
+  ) : null;
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/HolidayActionButton.tsx]: in such cases it's better to use `if` construction (easier to read)

File: webapp/packages/plugin-holidays/src/HolidaysService.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    private readonly topNavService: TopNavService,
+    private readonly administrationTopAppBarService: AdministrationTopAppBarService,
+  ) {
+    super();
+  }
+
+  override register() {
+    this.topNavService.placeholder.add(HolidayButton, 4);
+    this.administrationTopAppBarService.placeholder.add(HolidayButton, 4);
+  }
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/HolidaysService.ts]: same situation here, this service depends both on public and administration part, it needs to be split

File: webapp/packages/plugin-holidays/src/IHoliday.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+export interface IHoliday {
+  startDate: Date;
+  endDate: Date;
+  isHoliday: boolean;
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/IHoliday.ts]: `holiday.isHoliaday` is a bit weird, it may be active or ongoing or something else.

File: webapp/packages/plugin-holidays/src/IHoliday.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  name: string;
+  logoSrc: string;
+  iconSrc: string;
+  activeIconSrc?: string;
+  isCelebrating: boolean;
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/IHoliday.ts]: Now I'm baffled; what is the difference between `isHoliday`?
It will be better to name it `animationActive` and actions like `startAnimation`, `stopAnimation`
(or maybe another word, but something to make some meaning difference with `isHoliday`

File: webapp/packages/plugin-holidays/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { holidaysPlugin } from './manifest.js';
+
+export * from './HolidaysService.js';
+
+export default holidaysPlugin;
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/index.ts:14]: please also add named export

File: webapp/packages/plugin-top-app-bar/src/TopNavBar/TopNavService.ts [hunk 14]
```diff
@@ -9,15 +9,13 @@ import { Fill, PlaceholderContainer } from '@cloudbeaver/core-blocks';
...
 
 @injectable()
 export class TopNavService {
   readonly placeholder = new PlaceholderContainer();
 
   constructor() {
-    this.placeholder.add(Logo, 0);
     this.placeholder.add(MainMenu, 1);
     this.placeholder.add(Fill, 3);
     this.placeholder.add(AppStateMenu);
```
Comment (Wroud) [inline webapp/packages/plugin-top-app-bar/src/TopNavBar/TopNavService.ts:21]: please revert this

**Review by Wroud — CHANGES_REQUESTED** [review]

File: webapp/packages/plugin-app-logo-administration/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { appLogoAdministrationPlugin } from './manifest.js';
+
+export * from './PluginBootstrap.js';
```
Comment (devnaumov) [inline webapp/packages/plugin-app-logo-administration/src/index.ts]: Why do we need to export PluginBootstrap here?

File: webapp/packages/plugin-app-logo/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { appLogoPlugin } from './manifest.js';
+
+export * from './PluginBootstrap.js';
```
Comment (devnaumov) [inline webapp/packages/plugin-app-logo/src/index.ts]: Its enough to register bootstrap service in manifest file, no need to export it outside

File: webapp/packages/plugin-holidays/src/Christmas/Christmas.ts [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+  private lastFrameTime = 0;
+  private lastFlakeTime = 0;
+  private mX = -MIN_EFFECTIVE_DISTANCE;
+  private mY = -MIN_EFFECTIVE_DISTANCE;
+  private ctx: CanvasRenderingContext2D | null = null;
+  private canvas: HTMLCanvasElement | null = null;
+  private maxFlakesCount = 0;
+  private stopTimeoutId: number | undefined = undefined;
+  private mouseMovingTimeoutId: number | undefined = undefined;
+  private _isRunning = false;
```
Comment (devnaumov) [inline webapp/packages/plugin-holidays/src/Christmas/Christmas.ts]: Why only _isRunning has underscore? I think private mark is enough to understand that field is only used inside the class

File: webapp/packages/plugin-holidays/src/HolidayActionButton.tsx [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+      holiday?.startEffects();
+    }
+  }
+
+  if (!holiday) {
+    return null;
+  }
+
+  return (
+    <button
```
Comment (devnaumov) [inline webapp/packages/plugin-holidays/src/HolidayActionButton.tsx]: We have <IconButton />, ActionIconButton and IconOrImage. Can we use these components?

File: webapp/packages/plugin-holidays/src/HolidaysService.ts [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+  constructor(private readonly topNavService: TopNavService) {
+    super();
+  }
+
+  override register() {
+    this.topNavService.placeholder.add(HolidayButton, 4);
+  }
+
+  addHoliday(holiday: IHoliday) {
+    this.holidays.push(holiday);
```
Comment (devnaumov) [inline webapp/packages/plugin-holidays/src/HolidaysService.ts:30]: Please check whether holiday is already in array

File: webapp/packages/plugin-app-logo/src/Logo.tsx [hunk 20]
```diff
@@ -26,5 +28,5 @@ export const Logo = observer(function Logo() {
 
   const title = isSameVersion ? backendVersionTitle : commonVersionTitle;
 
-  return <AppLogo title={title} onClick={() => screenService.navigateToRoot()} />;
+  return <AppLogo title={title} iconSrc={holiday?.logoSrc ?? undefined} onClick={() => screenService.navigateToRoot()} />;
```
Comment (devnaumov) [inline webapp/packages/plugin-app-logo/src/Logo.tsx]: Just holiday?.logoSrc?

**Review by devnaumov — CHANGES_REQUESTED** [review]

File: webapp/packages/plugin-holidays/src/HolidayActionButton.tsx (hunk 18 above)
Comment (SychevAndrey) [reply webapp/packages/plugin-holidays/src/HolidayActionButton.tsx]: Used IconButton

File: webapp/packages/plugin-holidays/src/HolidayActionButton.module.css [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+button.holidayButton {
+  height: 48px;
+  width: 48px;
+  margin: 0;
+  padding: 12px;
+  display: flex;
+  justify-content: center;
```
Comment (devnaumov) [inline webapp/packages/plugin-holidays/src/HolidayActionButton.module.css]: I think most of the styles are already implemented on the IconButton level

File: webapp/packages/plugin-app-logo-administration/src/PluginBootstrap.ts [hunk 22]
```diff
@@ -7,18 +7,18 @@
  */
 import { Bootstrap, injectable } from '@cloudbeaver/core-di';
 import { AdministrationTopAppBarService, WizardTopAppBarService } from '@cloudbeaver/plugin-administration';
-import { Logo } from '@cloudbeaver/plugin-top-app-bar';
+import { Logo } from '@cloudbeaver/plugin-app-logo';
 
 @injectable()
-export class AdministrationTopAppBarBootstrap extends Bootstrap {
+export class PluginBootstrap extends Bootstrap {
```
Comment (Wroud) [inline webapp/packages/plugin-app-logo-administration/src/PluginBootstrap.ts]: it's better to keep name because it's really hard to debug problems in plugin bootstrap services (cuz it's a lot of them)

File: webapp/packages/plugin-holidays/package.json [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    "@cloudbeaver/core-di": "^0",
+    "@cloudbeaver/core-utils": "^0",
+    "@cloudbeaver/plugin-top-app-bar": "^0",
+    "mobx": "^6",
+    "mobx-react-lite": "^4",
+    "react": "^18"
+  },
+  "peerDependencies": {},
+  "devDependencies": {
+    "@types/react": "*",
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/package.json]: please specify version

File: webapp/packages/plugin-holidays/src/Christmas/Christmas.ts [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+    }, 100);
+  }, 60);
+
+  onResize = debounce(() => {
+    this.calculateMaxFlakesCount(window.innerWidth, window.innerHeight);
+    if (this.canvas) {
+      this.canvas.width = window.innerWidth;
+      this.canvas.height = window.innerHeight;
+    }
+  }, 500);
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/Christmas/Christmas.ts]: please use named functions in debounce and throttle

File: webapp/packages/plugin-holidays/src/Christmas/Christmas.ts [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+  onMouseMove = (e: MouseEvent) => {
+    this.handleMouseMove();
+    this.mX = e.clientX;
+    this.mY = e.clientY;
+  };
+
+  onMouseLeave = () => {
+    this.mX = -MIN_EFFECTIVE_DISTANCE;
+    this.mY = -MIN_EFFECTIVE_DISTANCE;
+  };
```
Comment (Wroud) [inline webapp/packages/plugin-holidays/src/Christmas/Christmas.ts]: please use `this.onMouseMove = this.onMouseMove.bind(this)` in the constructor instead or arrow function in the component

**Review by Wroud — CHANGES_REQUESTED** [review]
because of new logo plugins, we need PR in private repo as well

**Review by devnaumov — APPROVED** [review]

File: webapp/packages/plugin-app-logo/src/PluginBootstrap.ts [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { Bootstrap, injectable } from '@cloudbeaver/core-di';
+import { TopNavService } from '@cloudbeaver/plugin-top-app-bar';
+
+import { Logo } from './Logo.js';
```
Comment (Wroud) [inline webapp/packages/plugin-app-logo/src/PluginBootstrap.ts]: please use `importLazyComponent` - so main goal here is to initialize "business" logic as fast as possible and then load UI part so every time you import components in services use `importLazyComponent`

**Review by Wroud — APPROVED** [review]

**Comment by hadoopmarc** [comment]
Please help to disable this plugin. What should be in the configs?
